packages/context/src/context.ts

```diff
@@ -34,16 +52,142 @@ export class Context {
   protected _parent?: Context;
 
   /**
-   * Create a new context
+   * A list of registered context observers
+   */
+  protected readonly observers: Set<ContextObserver> = new Set();
+
+  /**
+   * Internal counter for pending events which observers have not processed yet
+   */
+  private pendingEvents = 0;
+
+  /**
+   * Create a new context. For example,
+   * ```ts
+   * // Create a new root context, let the framework to create a unique name
+   * const rootCtx = new Context();
+   *
+   * // Create a new child context inheriting bindings from `rootCtx`
+   * const childCtx = new Context(rootCtx);
+   *
+   * // Create another root context called "application"
+   * const appCtx = new Context('application');
+   *
+   * // Create a new child context called "request" and inheriting bindings
+   * // from `appCtx`
+   * const reqCtx = new Context(appCtx, 'request');
+   * ```
    * @param _parent The optional parent context
+   * @param name Name of the context, if not provided, a `uuid` will be
+   * generated as the name
    */
   constructor(_parent?: Context | string, name?: string) {
+    super();
     if (typeof _parent === 'string') {
       name = _parent;
       _parent = undefined;
     }
     this._parent = _parent;
     this.name = name || uuidv1();
+
+    this.setupEventHandlers();
+  }
+
+  /**
+   * Wrap the debug statement so that it always print out the context name
+   * as the prefix
+   * @param formatter
+   * @param args
+   */
+  // tslint:disable-next-line:no-any
+  private _debug(formatter: any, ...args: any[]) {
+    /* istanbul ignore if */
+    if (debug.enabled) {
+      if (typeof formatter === 'string') {
+        debug(`[%s] ${formatter}`, this.name, ...args);
+      } else {
+        debug('[%s] ', this.name, formatter, ...args);
+      }
+    }
+  }
+
+  /**
+   * Set up an internal listener to notify registered observers asynchronously
+   * upon `bind` and `unbind` events
+   */
+  private setupEventHandlers() {
+    // The following are two async functions. Returned promises are ignored as
+    // they are long-running background tasks.
+    this.startNotificationTask('bind').catch(err => {
+      // Catch error to avoid lint violations
```

Comment:
Please extract the error handler into a shared function to avoid duplication. The source code file `context.ts` is already way too long :(

Also this comment is no longer relevant, we are not catching the error only to avoid lint violations, we are forwarding it via `error` event.

```ts
const reportError = err => {
  debug(err);
  this.emit('error', err);
};

this.startNotificationTask('bind').catch(reportError);
this.startNotificationTask('unbind').catch(reportError);
```

Reply:
Fixed.